### ISSUE TITLE
Code generation: how to avoid naming collisions

### ISSUE THREAD

**Issue by lmolkova** [open-telemetry/semantic-conventions#1118]
As we discovered in #1031, certain semantic conventions changes (together with default code generator behavior) result in ambiguous constant names generated in the code.

For example, when `foo.bar_baz` is renamed to `foo.bar.baz`, the code generator produces the same constant name for both -`FOO_BAR_BAR` , `FooBarBaz`, etc depending on the preferred casing.

We don't remove attributes/enum members/metrics from the semantic conventions anymore (old property is deprecated), so both constants would exist in the same version.

So, the semantic conventions together with the codegen should prevent such collisions from happening:
- The default tooling behavior should result in collision-free code.
- There could be additional policies added to semconv to prevent certain renames.

**Comment by lmolkova** [comment]
Option 1:

- any renames are allowed (for experimental conventions)
- `_` is preserved in constant names:
  - `foo.bar_baz` is generated as `FOO_BAR__BAZ`, `FooBar_Baz`, etc

Cons:
- `__` is ugly in snake_case and `_` is ugly in PascalCase

**Comment by lmolkova** [comment]
Option 1.5:

`foo.bar_baz` is generated as `FOO__BAR_BAZ`, `Foo_BarBaz`, ...

**Comment by lmolkova** [comment]
Option 2:

`foo.bar_baz` is generated as `FOO_BARBAZ`, `FooBarbaz`, ...

- renames that only remove or add `_` are not allowed even for experimental attributes. 
- `_` is removed during code generation

Cons:
- limited rename options. Partially mitigated by the fact that we didn't recall making such renames in the past.

**Comment by lmolkova** [comment]
Option 3 (do nothing):

- any renames are allowed (for experimental conventions)
- `_` and `.` are not distinguishable in constant names

Cons:

- attribute name and schema version mismatch (instrumentation thinks it sets `foo.bar_baz`, but it sets `foo.bar.baz`.

**Comment by lmolkova** [comment]
Option 4:

`foo.bar_baz` is generated as `FOO_BAR_BAZ`, `FooBarBaz`.
When `foo.bar.baz` is added, the collision is detected, so the new attribute is called `FOO_BAR_BAZ_NEW`, `FooBarBazNew`.

- any renames are allowed (for experimental conventions)
- resolve collisions by giving new names 

Cons:

- new attribute would have a 'bad' name and deprecated attribute would have a `good` name forever.  

The rename can be done for the old attribute, resulting in Option 3.

**Comment by marcalff** [comment]
When 2 different symbols in semantic conventions generate the same symbol on the generated code,
that is, when there is a collision, one of the name mapping has to change, and this change is by definition a breaking change: code instrumented that refers to a given name has to be adjusted to use the adjusted name instead.

To minimize the overall impact, it is preferable that only symbols that are the least frequent gets to be broken.

Any solution that changes the output for `foo.bar` to be different, which is the most common case, should be discarded.

So, solution 1.5 is definitively out in my opinion, it will break everything.

Assuming semantic conventions containing the `_` character are relatively less represented (in all the semconv that exists), the change should be on how `_` is represented, to minimize the overall impact, without touching `.`.

Option 2 is out, it collides on `foo.bar_baz` and `foo.barbaz`.

Option 4 does not scale in my opinion, after several renames, and it assumes there is some "state" stored somewhere to remember if a collision ever existed in the past or not. Does not sound viable.

Option 3 does not resolve the issue, this is bound to be repeated. For new semantic conventions, a desirable goal is to be able to decide the most natural name for a given semconv, without having to look at possible collisions, that would prevent to use a good name ... especially when the colliding semconv is deprecated and will eventually be removed over time.

The only viable solution looks like option 1.

**Comment by goedelsoup-clari** [comment]
What's the problem with the final descriptor winning? This defines a predictable behavior which anyone can bypass by defining stricter behaviors backwards in their models.

As a Haskeller, I expect naming collisions and a stronger identity system here, not a resolution approach. This should be concrete in the implementation and type resolution IMO.

**Comment by austinlparker** [comment]
> ### Option 2: `foo.bar_baz` is generated as `FOO_BARBAZ`, `FooBarbaz`, ...
> * **renames that only remove or add `_` are not allowed even for experimental attributes.**
> * `_` is removed during code generation
> 
> Cons:
> 
> * limited rename options. Partially mitigated by the fact that we didn't recall making such renames in the past.

just to check, does that mean that `messaging.client_id` -> `messaging.client.id` gets rolled back and there's no change?

**Comment by lmolkova** [comment]
According to the Option2, there will be no collision for `messaging.client_id` -> `messaging.client.id`:
-  `messaging.client_id` would be `MESSAGING_CLIENTID` or `MessagingClientid`
- `messaging.client.id` would be `MESSAGING_CLIENT_ID` or `MessagingClientId`

So I don't think we need to roll it back.

> Option 2 seems the preferable approach, but just to confirm this means that any semconv generated code after this is implemented will cause a breaking change with previous versions as the variables will have been renamed, correct?

That's correct. Any option we pick (except opt3) will result in breaking changes in generated code. In most languages this is tolerable since semconv artifact is experimental.
So far we've identified JS and PHP which shipped stable semconv artifacts and JS was going to make some breaking changes anyway. 
It will be a problem for PHP.

**Comment by marcalff** [comment]
> Another point that in this specific case: The new attribute `messaging.client.id` needs to produce a breaking change. 

It will not.

Existing code using `MESSAGING_CLIENT_ID` or `MessagingClientId`, corresponding to `messaging.client_id`, will need to be fixed to use `MESSAGING_CLIENTID` or `MessagingClientid` instead, once code generation is fixed.

**Comment by joaopgrassi** [comment]
Ah yes, that for sure. I think I didn't do a good job explaining, but what I was after is: whatever we do, the old const name **can't carry the new attribute name**. Because as you said, that wouldn't break, and be even worse because it would start sending the new attribute name.

**Comment by marcalff** [comment]
See my previous comment:

> When 2 different symbols in semantic conventions generate the same symbol on the generated code,
> that is, when there is a collision, one of the name mapping has to change, and this change is by definition a breaking change: code instrumented that refers to a given name has to be adjusted to use the adjusted name instead.
> 

I don't think there is any way around this.

Failure to update the name will use the new semantic convention, so using `MESSAGING_CLIENT_ID` or `MessagingClientId` will mean `messaging.client.id` while it meant `messaging.client_id` before.

**Comment by joaopgrassi** [comment]
You're right, seems I also misunderstood option 2. I somehow thought the new one would be changed and not be used automatically, I see it now. 

I think this is rather bad :(. It will be hard to not overlook something and end up with this wrong situation of old instrumentation producing/using the new attributes while not fully yet converted to the stable conventions. 

The only way to not have this, would be to break both old and new generated consts. But breaking the mapping of `.` to `_` in code gen is also bad.

**Comment by dyladan** [comment]
> So far we've identified JS and PHP which shipped stable semconv artifacts and JS was going to make some breaking changes anyway.

In JS we're just marking everything currently exported by the package as deprecated and keeping it until we go to 2.0. We're exporting all the new names with a new style anyway to be friendlier to tree shakers and code minifiers (important in JS) so any breaking name changes due to a change in code generation are fine. If it happens again we'll just rev the major version as this package is entirely separate from the rest of the JS client.

In short: option 2 is fine for JS and is my preference

**Comment by dyladan** [comment]
Here's what the option 2 codegen looks like in JS https://github.com/open-telemetry/opentelemetry-js/pull/4690/commits/dbb8328ddc072adf5bd32768fe28529328135f53

note: we aren't merging any changes until this issue is resolved, this is just to prototype and make sure we're ready to move quickly when a final decision is made.

**Comment by lmolkova** [comment]
There is an overwhelming support for option 2:
- `_` should not be rendered when generating constants (constants, fields, class names, method names, files names, etc)
- semconv will have a policy to prevent renames that only add or remove underscores.

We're working on the tooling updates and have a couple of draft PRs which show how things would looks like:
- Java: https://github.com/open-telemetry/semantic-conventions-java/pull/75
- Python: https://github.com/open-telemetry/opentelemetry-python/pull/3927

Before we ship tooling update would like to get any last-minute feedback from everyone interested, specifically @lzchen, @jack-berg, @trask on the above PRs.

**Comment by jack-berg** [comment]
Discussed at today's Java SIG and option 2 was the preference. Thanks!

**Comment by lmolkova** [comment]
Option 2 does not look good in the real life
- `DB_CASSANDRA_CONSISTENCYLEVEL`
- `DbCassandraConsistencylevelValues`
- `AWSELASTICBEANSTALK`

https://github.com/open-telemetry/semantic-conventions-java/pull/75

`AWS__ELASTIC__BEAN__STALK`, etc would be more readable. 

I'd like to go back and explore Option 1 (`__`) or  Option 3 (do nothing - `_` and `.` are the same in the code).

Option 3+ may look like:
- non-deprecated attribute get the constant name, value is updated. The deprecated one is not generated - this could be tolerable for experimental conventions
- semconv policy would prevent 2 non-deprecated attributes to have the same const name.
- we won't allow to rename stable attributes (without major version bump) anyway

We can also prohibit `.` <-> `_` renames for all attributes including experimental.

**Comment by lmolkova** [comment]
Here's more details on the Option 3.5:
- allow collisions between deprecated and non-deprecated attributes, ignoring deprecated.
- not allow collisions between non-deprecated attributes.

Pros:
- we generate the same code, no breaking changes 
- the code is **not ugly** (neither `AWS__ELASTIC__BEAN__STALK` nor `AWSELASTICBEANSTALK`)
- we have a policy to prevent collisions for non-deprecated attributes
- we can still clean up experimental attributes if we want to rename it
- it won't affect stable attributes - we should not rename them

Cons:
- possible attribute name and schema version mismatch (instrumentation thinks it sets foo.bar_baz, but it's foo.bar.baz now).

Prototypes:
- Java example: https://github.com/open-telemetry/semantic-conventions-java/pull/75
- Build-tools policy + codegen  impl: https://github.com/open-telemetry/build-tools/pull/324
- Weaver codegen example is on the way

**Comment by dyladan** [comment]
I think option 3.5 with prohibiting `.` <-> `_` for all including experimental would be my preferred solution as it generates the nicest code and retains compatibility with the current generation. I think it is important to include that restriction in order to avoid telemetry shape changing without user intervention and the mentioned name and schema version mismatch.

The biggest question IMO would be then what to do about the collisions that have already happened. We can probably just accept the renames that have already happened as an accident of history.

**Comment by trisch-me** [comment]
i'm feeling prohibiting `.` <-> `_` will limit our option to extract a namespace out of existing attributes in the future.

consider `x.user_id` and `x.user_name` which should become `x.user.id` and `x.user.name`

**Comment by trask** [comment]
I'll take this (back) to the Java SIG on Thursday.

I suspect that we may prefer this trade off:
* keep the nice readable constant names
* allow breaking changes in our experimental semconv artifact when there is a conflicting rename (e.g. `foo.bar_baz` &rarr; `foo.bar.baz`)

note that Java already [recommends that libraries make copies of experimental attributes](https://github.com/open-telemetry/semantic-conventions-java/tree/main#published-released) in order to avoid the diamond dependency problem:

> Generated code for experimental semantic conventions.</br>**NOTE:** This artifact has the [-alpha](https://github.com/open-telemetry/opentelemetry-java/blob/main/VERSIONING.md) and comes with no compatibility guarantees. Libraries can use this for testing, but should make copies of the attributes to avoid possible runtime errors from version conflicts.

see https://github.com/open-telemetry/semantic-conventions-java/issues/50#issuecomment-1998332270 for a bit more background on this recommendation:

> - `opentelemetry-semconv-incubating` contains stable and incubating semantic conventions
>   - Classes live in `io.opentelemetry.semconv.incubating` package. Note this is different than stable artifact to make java module system work.
>   - Stable attributes will be annotated `@deprecated` with javadoc pointing to the equivalent location in the stable artifact
>   - The `semantic-conventions` repo will keep incubating attributes around indefinitely (or for a long time) and mark as deprecated. These deprecated, experimental attributes will be annotated `@deprecated`.
>   - This artifact will always be marked `alpha`, since we may eventually remove some deprecated experimental attributes, and since the types of experimental attributes are subject to change.
>   - We will discourage library authors from using this artifact. They can use it for making assertions in tests, but should make copies of any experimental attributes they depend on in instrumentation.

**Comment by lmolkova** [comment]
Discussed at Semconv and Maintainers calls:
- there is a trade-off between readable/idiomatic code and collisions
- the collisions affect experimental attributes only. we'll not rename stable attributes
- languages/SIGs should be able to generate (or not generate) both constants (i.e. https://github.com/open-telemetry/semantic-conventions/issues/1118#issuecomment-2158731384 - removing an attribute from semconv is not an option)
- we probably have other candidates for `_` <-> `.` renames (I'll share an update), if we restrict such renames now, we'll never have a chance to fix things.

One of the possible solutions includes a grace period during which `_` <-> `.` are allowed.

**Comment by marcalff** [comment]
Changing code generation rules in general for `_` will affect a lot of existing, non deprecated, non colliding semantic conventions, which is not desirable.

How about:

* Define a new property, `generate_as`, in build tools / weaver
* In semantic conventions, **_manually resolve collisions_** by providing an alternate `generate_as` name for the generated symbol, while not changing the associated value

For example:

* semconv `messaging.client.id` is untouched.
* semconv `messaging.client_id` is modified with the property `generate_as = messaging.client_id.deprecated` (since, after all, it is deprecated. Could be named `.old`, `.v1`, or anything different)

Taking C++ as an example, code generation will look like:

```
  kMessagingClientId = "messaging.client.id"; // new semconv
  kMessagingClientIdDeprecated = "messaging.client_id"; // old semconv.
```

The major benefit I see is that existing semconv like:
  * `k8s.container.restart_count`
  * etc

are unchanged.

Only symbol that **_actually collide_** need special treatment, and only user code that depends on these symbols need to change (to use kMessagingClientIdDeprecated if this is really what the instrumentation wants).

What `generate_as` provides, is to decouple symbol names from semantic conventions values, so resolving a conflict in values does not force to revise code generation rules names for everything, impacting many unrelated symbols (all the existing semconv that contains a `_` today).

**Comment by marcalff** [comment]
@lmolkova Please see proposal for option 5 above.

**Comment by lmolkova** [comment]
@marcalff thanks! 

It does not make send to generate the deprecated attribute differently - nobody will update their code to point to a deprecated attribute - we don't want them to.

I.e. having `generate_as` would only make sense on the new attribute. But also we don't want to change how the new attribute is generated in the stable conventions.

But I really like the idea of manual resolution:
- the collisions would be rare
- linters would detect duplicates, I.e. it'd be quite hard to ship the artifact without first resolving collisions
- the manual resolution would be one of the following:
  - drop the deprecated attribute - some languages could be fine with it.
  - give a new name to the new attribute. 
    - E.g. `messaging.client.id` would become `incubating-semconv.MessagingAttributes.MESSAGING_CLIENT_ID2 = "messaging.client.id"` and `stable-semconv.MessagingAttributes.MESSAGING_CLIENT_ID = "messaging.client.id"`
  - someone can still do Option 5 and give a new name to OLD attribute
    - E.g. `messaging.client_id` would become `incubating-semconv.MessagingAttributes.MESSAGING_CLIENT_ID_DEPRECATED = "messaging.client_id"`

More context on this here:
I think @trask was also suggesting it here https://github.com/open-telemetry/semantic-conventions-java/pull/75#discussion_r1630446636

and here's the prototype:  https://github.com/open-telemetry/semantic-conventions-java/pull/76 - effectively there is a way to remap constant name and/or remove attribute.

**Comment by marcalff** [comment]
> TL;DR: **SIGs decide to drop the deprecated attribute or assign a different constant name - they pick a new constant name in case of a collision.**

I don't understand.

I was hopping for a solution where the alternate name comes from metadata added in the semantic convention for collisions, so that the code generation scripts do not have to be changed all the time to account for special cases as they arise.

Now it looks like each SIG, in their code generation scripts, it to add code for special cases.

Is the conclusion really that SIGs are on their own and must fix collisions created "upstream" in semantic conventions ?

In this case, opentelemetry-cpp already complies: it drops the deprecated `messaging.client_id`.

**Comment by lmolkova** [comment]
> I was hopping for a solution where the alternate name comes from metadata added in the semantic convention for collisions, so that the code generation scripts do not have to be changed all the time to account for special cases as they arise.

Assuming there is an alternative constant name in the schema, would you drop the attribute or would you generate both attributes  (with different names) in C++? 

The prescriptive solution would be:
- `messaging.client_id` (old) should be marked so it can be dropped (if that's the SIG choice). it. E.g. `collision: messaging.client.id`
- `messaging.client.id` (new) would be marked with alternative constant name. E.g. `unque_const_name: messaging_client_id2`.

Jinja scripts would need to be changed anyway to either drop or use constant name  from the `unque_const_name` (converting it to  proper case).

> Now it looks like each SIG, in their code generation scripts, it to add code for special cases.

The templates need to be changed now to support dropping/renaming. If a collision happens again, the change would be to add another attribute name to the list of excluded attributes (or to a manually maintained map). E.g. this change is java https://github.com/open-telemetry/semantic-conventions-java/pull/76 adds all you need to either drop or rename constants by making a small config change.

> Is the conclusion really that SIGs are on their own and must fix collisions created "upstream" in semantic conventions ?

I think it's a path forward and not a conclusion. 
We need need to add checks to semconv to know when they happen before we release. We should probably block such renames until we have a reasonable conclusion.

Manual resolution seems like an easy and cheap thing to do to solve this and similar collisions in the same way.  We can automate it, but it's still not clear to me what are all the options we need to provide.

Note:
We are migrating tooling from build-tools to weaver.  Code generation migration which would involve some jinja updates and this would be a good time to add the `final' solution. 

TL;DR:
- semconv should detect codegen collisions when they are added
- I want to see how SIGs handle the current collision to understand which options we need to provide
- while we're discussing this, we should block renames that cause such collisions

**Comment by lmolkova** [comment]
Discussed at Semconv WG 6/17

- we need to detect codegen collisions when they are added to semconv
- we should put a moratorium on renames that'd result in collision until we have tooling/checks in place
- we need to update attribute naming guidance and suggest to use `.` when it makes sense

- we'd probably make different choices for different attributes
  - in case of `messaging.client_id` dropping the deprecated/experimental attribute is reasonable
  - we might want different solution for `process.command_line` 
  - semconv should be the source of guidance on dropping/renaming
  - we need tooling for it

So:
- phase 1:
  - add collision checks to semconv tooling
  - enforce moratorium on collisions
- phase 2:
  - tooling work to support dropping/remapping based on the schema
- phase 3:
  - lift moratorium and allow colliding renames
  - rename things we need to rename
- phase 4:
  - eventually prohibit  renames that result in collisions

**Comment by dyladan** [comment]
Has there been any resolution on this?

**Comment by lmolkova** [comment]
Yes, we're adding collision checks - https://github.com/open-telemetry/semantic-conventions/pull/1209 - necessary tooling changes just landed.
